### PR TITLE
refactor: extract LRU package

### PR DESCRIPTION
**Pull request by rhybrillou**
## Description

In order to prepare for the changes to the rate limited logger, a specific kind of LRU cache has to be added.
The point of this refactor is to have an abstraction for LRU caches and add a LRU cache with item expiration.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

CI should be enough.
A benchmark as well as a set of tests has been added to the new ExpirableLRU type.

### CODE REVIEW

**Comment by rhybrillou** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6793** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6793" target="_blank">Graphite</a>  👈
    * **PR #5854** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5854" target="_blank">Graphite</a> 
      * **PR #6195** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6195" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6793?utm_source=stack-comment).

File: pkg/lru/cache.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package lru
```
Comment (janisz) [inline pkg/lru/cache.go:1]: nit: I think the standard in our repo is to have `cache.go` for interface and `cache_impl.go` for implementation

File: pkg/lru/expirable_lru.go [hunk 2]
```diff
@@ -0,0 +1,463 @@
+package lru
+
+/*
+source: https://github.com/hashicorp/golang-lru/pull/116/files#diff-be4b9429f63e821595d57501426a7f91505ab8f2df186c5f5e3b0af74e1a5dfc
```
Comment (janisz) [inline pkg/lru/expirable_lru.go:4]: Please add `// TODO(ROX-???):`

File: pkg/lru/list.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE_list file.
+
+package lru
+
+import "time"
+
+/*
+Source: https://github.com/hashicorp/golang-lru/blob/8d9a62dcf60cd87ed918b57afad8a001d25db3de/simplelru/list.go
```
Comment (janisz) [inline pkg/lru/list.go:10]: Maybe we should fork and replace this dep and not copy code directly?

File: pkg/lru/test_expirable_lru.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
+package lru
+
+import (
+	"testing"
+	"time"
+)
+
```
Comment (janisz) [inline pkg/lru/test_expirable_lru.go:7]: Is this test copied or created from scratch?

Comment (rhybrillou) [reply pkg/lru/test_expirable_lru.go:7]: This file is a utility for testing purposes that was created from scratch.

File: pkg/grpc/metrics/grpc_metrics_impl.go [hunk 5]
```diff
@@ -4,9 +4,9 @@ import (
 	"context"
 	"runtime/debug"
 
-	lru "github.com/hashicorp/golang-lru/v2"
 	"github.com/pkg/errors"
 	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/lru"
```
Comment (mtodor) [inline pkg/grpc/metrics/grpc_metrics_impl.go:9]: I would suggest minimizing the usage of our `lru` package. It's better to use Hashicorp LRU unless we have to use our own. And that should be only in the case of ExpriableLRU.

This applies to all 4 files where we did such changes.

File: pkg/lru/cache_impl.go
Comment (mtodor) [inline pkg/lru/cache_impl.go:1]: I would suggest adding only NewExpirableLRU. For all other caches, we should still use caches provided by the caching library.

File: pkg/logging/internal/lru/expirable_lru_test.go
Comment (mtodor) [inline pkg/logging/internal/lru/expirable_lru_test.go:1]: I would suggest adding two small tests.

1. That key is evicted when add is called for full cache
```
func TestExpirableLRUAddWithEvict(t *testing.T) {
	var evicted int
	lc := NewExpirableLRU(1, func(_, _ string) { evicted++ }, 0)

	lc.Add("t1", "t1")
	lc.Add("t1", "t2")
	if evicted != 0 {
		t.Fatalf("adding of the same key should not evict")
	}

	lc.Add("t2", "t1")
	if evicted != 1 {
		t.Fatalf("add new key should evict old key if there is no space")
	}
}
```

2. When we reduce size of full cache, that evict callback is triggered
```
func TestExpirableLRUResizeWithEvict(t *testing.T) {
	var evicted int
	lc := NewExpirableLRU(2, func(_, _ string) { evicted++ }, 0)

	lc.Add("t1", "t1")
	lc.Add("t2", "t2")
	if evicted != 0 {
		t.Fatalf("cache is not full and there should be no evict")
	}

	lc.Resize(1)
	if evicted != 1 {
		t.Fatalf("reducing size should evict")
	}
}
```

**Review by mtodor — COMMENTED** [review]
This looks great! I have added two minor comments.

File: pkg/logging/internal/lru/cache_impl.go
[the remark this answers is not in the document]
Comment (rhybrillou) [reply pkg/logging/internal/lru/cache_impl.go:1]: It is a leftover of the time when I wanted to have a package that wraps around the external LRU component.
I'll remove it.

**Review by mtodor — APPROVED** [review]
Nice!!!

**Pull request merged**